File: src/shared/fingerprint-args.ts
/**
 * Translate a Hub profile into CloakBrowser launch options.
 *
 * Pure functions only — no Node APIs — so this module is unit-testable and can
 * also be imported by the renderer to *preview* the exact flags a profile will
 * launch with (the "Advanced → resolved flags" panel).
 *
 * Flag reference (CloakBrowser binary):
 *   --fingerprint=<seed>                master seed: canvas/WebGL/audio/fonts/rects
 *   --fingerprint-platform=<os>         navigator.platform, UA OS, GPU pool
 *   --fingerprint-platform-version=     Client Hints platform version
 *   --fingerprint-brand / -brand-version
 *   --fingerprint-screen-width/-height
 *   --fingerprint-gpu-vendor/-renderer
 *   --fingerprint-hardware-concurrency  navigator.hardwareConcurrency
 *   --fingerprint-device-memory         navigator.deviceMemory (GB)
 *   --fingerprint-storage-quota         MB — raise to look non-incognito
 *   --fingerprint-timezone / -locale
 *   --fingerprint-location=<lat,lon>
 *   --fingerprint-noise=false           disable noise, keep deterministic seed
 *   --fingerprint-windows-font-metrics  Chromium 148+
 *   --fingerprint-allow-3p-cookies      Chromium 148+
 *   --fingerprint-fonts-dir=<path>
 *   --fingerprint-webrtc-ip=auto|<ip>
 *   --fingerprint-taskbar-height=<px>
 */

import type { Profile } from './types';

/** Flags the app always owns — a user-supplied duplicate is overridden. */
const OWNED_PREFIXES = ['--fingerprint', '--lang'];

export interface ResolvedLaunch {
  /** Chromium CLI flags (deduplicated, deterministic order). */
  args: string[];
  /** Value for the wrapper's `timezone` option, if pinned. */
  timezone?: string;
  /** Value for the wrapper's `locale` option, if pinned. */
  locale?: string;
  /** Ask the wrapper to resolve timezone/locale from the proxy exit IP. */
  geoip: boolean;
  headless: boolean;
  humanize: boolean;
  humanPreset: 'default' | 'careful';
  humanConfig?: Record<string, unknown>;
  extensionPaths: string[];
  userAgent?: string;
  proxy?: string | { server: string; bypass?: string; username?: string; password?: string };
}

/** Build a proxy value in the shape the CloakBrowser wrapper expects. */
export function buildProxyOption(profile: Profile): ResolvedLaunch['proxy'] {
  const p = profile.proxy;
  if (!p || p.kind === 'none' || !p.host || !p.port) return undefined;
  const scheme = p.kind === 'socks5' ? 'socks5' : p.kind;
  const server = `${scheme}://${p.host}:${p.port}`;
  // Object form keeps credentials out of the server string, which avoids
  // double-encoding problems with passwords containing @ : / etc.
  const out: { server: string; bypass?: string; username?: string; password?: string } = { server };
  if (p.username) out.username = p.username;
  if (p.password) out.password = p.password;
  if (p.bypass) out.bypass = p.bypass;
  return out;
}

/** Human-readable proxy string for the UI (password masked). */
export function proxyLabel(profile: Profile): string {
  const p = profile.proxy;
  if (!p || p.kind === 'none' || !p.host) return 'Direct (no proxy)';
  const auth = p.username ? `${p.username}:••••@` : '';
  return `${p.kind}://${auth}${p.host}:${p.port ?? ''}`;
}

/**
 * Build the fingerprint-related Chromium flags for a profile.
 *
 * Only values the user pinned (`mode: 'manual'`) become explicit flags; auto
 * values are deliberately left to the binary so they stay coherent with the
 * seed. `extraArgs` are applied last and win on conflict, except for flags the
 * app owns (`--fingerprint*`, `--lang`) where the profile wins — otherwise a
 * stray user flag could silently break the identity the profile promises.
 */
export function buildFingerprintArgs(profile: Profile): string[] {
  const fp = profile.fingerprint;
  const flags = new Map<string, string>();
  const set = (flag: string, value?: string | number | boolean) => {
    flags.set(flag, value === undefined ? flag : `${flag}=${value}`);
  };

  if (typeof fp.seed === 'number' && Number.isFinite(fp.seed) && fp.seed > 0) {
    set('--fingerprint', Math.floor(fp.seed));
  }
  set('--fingerprint-platform', fp.platform);
  if (fp.platformVersion) set('--fingerprint-platform-version', fp.platformVersion);
  if (fp.brand && fp.brand !== 'Chrome') set('--fingerprint-brand', fp.brand);
  if (fp.brandVersion) set('--fingerprint-brand-version', fp.brandVersion);

  if (fp.screen.mode === 'manual' && fp.screen.width && fp.screen.height) {
    set('--fingerprint-screen-width', fp.screen.width);
    set('--fingerprint-screen-height', fp.screen.height);
  }
  if (fp.gpu.mode === 'manual') {
    if (fp.gpu.vendor) set('--fingerprint-gpu-vendor', fp.gpu.vendor);
    if (fp.gpu.renderer) set('--fingerprint-gpu-renderer', fp.gpu.renderer);
  }
  if (fp.cpuCores.mode === 'manual' && fp.cpuCores.value) {
    set('--fingerprint-hardware-concurrency', fp.cpuCores.value);
  }
  if (fp.deviceMemory.mode === 'manual' && fp.deviceMemory.value) {
    set('--fingerprint-device-memory', fp.deviceMemory.value);
  }
  if (typeof fp.storageQuotaMb === 'number' && fp.storageQuotaMb > 0) {
    set('--fingerprint-storage-quota', Math.floor(fp.storageQuotaMb));
  }
  if (typeof fp.taskbarHeight === 'number' && fp.taskbarHeight >= 0) {
    set('--fingerprint-taskbar-height', Math.floor(fp.taskbarHeight));
  }
  // Noise is ON by default in the binary; only the opt-out needs a flag.
  if (fp.noise === false) set('--fingerprint-noise', 'false');
  if (fp.windowsFontMetrics && fp.platform === 'windows') {
    set('--fingerprint-windows-font-metrics');
  }
  if (fp.fontsDir) set('--fingerprint-fonts-dir', fp.fontsDir);
  if (fp.allowThirdPartyCookies) set('--fingerprint-allow-3p-cookies');

  // WebRTC: only meaningful behind a proxy — a spoofed ICE IP with a direct
  // connection is itself a mismatch, so 'auto' is skipped without a proxy.
  const hasProxy = profile.proxy?.kind && profile.proxy.kind !== 'none' && !!profile.proxy.host;
  if (fp.webrtc.mode === 'manual' && fp.webrtc.ip) {
    set('--fingerprint-webrtc-ip', fp.webrtc.ip);
  } else if (fp.webrtc.mode === 'auto' && hasProxy) {
    set('--fingerprint-webrtc-ip', 'auto');
  }

  // Geolocation: explicit coordinates only. 'ip' mode is handled by geoip in
  // the wrapper, 'off' leaves the binary default untouched.
  if (profile.geo?.mode === 'manual' && profile.geo.latitude != null && profile.geo.longitude != null) {
    set('--fingerprint-location', `${profile.geo.latitude},${profile.geo.longitude}`);
  }

  // Pinned locale must also reach --lang so Accept-Language matches.
  if (profile.locale?.mode === 'manual' && profile.locale.locale) {
    set('--lang', profile.locale.locale);
    set('--fingerprint-locale', profile.locale.locale);
  }
  if (profile.locale?.mode === 'manual' && profile.locale.timezone) {
    set('--fingerprint-timezone', profile.locale.timezone);
  }

  // User extra args last: they may add flags we don't model, but they may not
  // hijack the identity flags above.
  for (const raw of profile.startup?.extraArgs ?? []) {
    const arg = raw.trim();
    if (!arg.startsWith('--')) continue;
    const key = arg.split('=')[0]!;
    const owned = OWNED_PREFIXES.some((p) => key === p || key.startsWith(p + '-'));
    if (owned && flags.has(key)) continue;
    flags.set(key, arg);
  }

  return [...flags.values()];
}

/** Build the full set of options for `launchPersistentContext()`. */
export function resolveLaunch(profile: Profile): ResolvedLaunch {
  const usesIpLocale = profile.locale?.mode === 'ip';
  const hasProxy = !!(profile.proxy?.kind && profile.proxy.kind !== 'none' && profile.proxy.host);

  const humanConfig: Record<string, unknown> = {};
  if (typeof profile.behaviour?.mistypeChance === 'number') {
    humanConfig.mistype_chance = profile.behaviour.mistypeChance;
  }
  if (typeof profile.behaviour?.typingDelay === 'number') {
    humanConfig.typing_delay = profile.behaviour.typingDelay;
  }
  if (profile.behaviour?.idleBetweenActions) {
    humanConfig.idle_between_actions = true;
  }

  return {
    args: buildFingerprintArgs(profile),
    // geoip only makes sense behind a proxy; without one the wrapper would
    // resolve the host's own IP, which is not what "follow the proxy" means.
    geoip: usesIpLocale && hasProxy,
    timezone: profile.locale?.mode === 'manual' ? profile.locale.timezone : undefined,
    locale: profile.locale?.mode === 'manual' ? profile.locale.locale : undefined,
    headless: profile.startup?.headless ?? false,
    humanize: profile.behaviour?.humanize ?? false,
    humanPreset: profile.behaviour?.preset ?? 'default',
    humanConfig: Object.keys(humanConfig).length ? humanConfig : undefined,
    extensionPaths: profile.startup?.extensionPaths ?? [],
    userAgent: profile.userAgent || undefined,
    proxy: buildProxyOption(profile),
  };
}

File: dotnet/src/CloakHub.Core/Launch/FingerprintArgs.cs
using System.Globalization;
using CloakHub.Core.Model;

namespace CloakHub.Core.Launch;

/// <summary>
/// Translates a Hub profile into CloakBrowser launch flags.
/// <para>
/// Pure functions only — no file, process or network access — so this is
/// unit-testable and the UI can call it to <i>preview</i> the exact argv a
/// profile will launch with. That preview is not a nicety: three of the bugs
/// fixed in the Electron version were diagnosed by reading it.
/// </para>
/// <para>Flag reference (CloakBrowser binary):</para>
/// <code>
///   --fingerprint=&lt;seed&gt;               master seed: canvas/WebGL/audio/fonts/rects
///   --fingerprint-platform=&lt;os&gt;         navigator.platform, UA OS, GPU pool
///   --fingerprint-platform-version=     Client Hints platform version
///   --fingerprint-brand / -brand-version
///   --fingerprint-screen-width/-height
///   --fingerprint-gpu-vendor/-renderer
///   --fingerprint-hardware-concurrency  navigator.hardwareConcurrency
///   --fingerprint-device-memory         navigator.deviceMemory (GB)
///   --fingerprint-storage-quota         MB — raise to look non-incognito
///   --fingerprint-timezone / -locale
///   --fingerprint-location=&lt;lat,lon&gt;
///   --fingerprint-noise=false           disable noise, keep deterministic seed
///   --fingerprint-windows-font-metrics  Chromium 148+
///   --fingerprint-allow-3p-cookies      Chromium 148+
///   --fingerprint-fonts-dir=&lt;path&gt;
///   --fingerprint-webrtc-ip=auto|&lt;ip&gt;
///   --fingerprint-taskbar-height=&lt;px&gt;
/// </code>
/// </summary>
public static class FingerprintArgs
{
    /// <summary>Flag prefixes the app owns — a user-supplied duplicate is ignored.</summary>
    private static readonly string[] OwnedPrefixes = ["--fingerprint", "--lang"];

    /// <summary>
    /// Stable fallback seed for a profile whose seed field was left empty.
    /// <para>
    /// FNV-1a over the profile id, mapped into the 10000-99999 range the wrapper
    /// itself uses. Deterministic on purpose: the same profile keeps the same
    /// device identity across launches and app restarts without persisting
    /// anything, and the flag preview stays stable between renders.
    /// </para>
    /// </summary>
    public static int SeedFromId(string id)
    {
        unchecked
        {
            uint h = 0x811c9dc5;
            foreach (var ch in id)
            {
                h ^= ch;
                h *= 0x01000193;
            }
            return 10000 + (int)(h % 90000);
        }
    }

    private static string PlatformFlag(FingerprintPlatform p) => p switch
    {
        FingerprintPlatform.Windows => "windows",
        FingerprintPlatform.Macos => "macos",
        FingerprintPlatform.Linux => "linux",
        _ => "windows",
    };

    /// <summary>
    /// Build the fingerprint-related Chromium flags for a profile.
    /// <para>
    /// Only values the user pinned (<c>Manual</c>) become explicit flags; auto
    /// values are left to the binary so they stay coherent with the seed.
    /// <c>ExtraArgs</c> are applied last and win on conflict, <i>except</i> for
    /// flags the app owns, where the profile wins — otherwise a stray user flag
    /// could silently break the identity the profile promises.
    /// </para>
    /// </summary>
    public static List<string> Build(Profile profile)
    {
        var fp = profile.Fingerprint;
        var flags = new Dictionary<string, string>(StringComparer.Ordinal);

        void Set(string flag, string? value = null) =>
            flags[flag] = value is null ? flag : $"{flag}={value}";

        void SetInt(string flag, int value) =>
            Set(flag, value.ToString(CultureInfo.InvariantCulture));

        // ------------------------------------------------------------------
        // A seed is ALWAYS emitted, never omitted.
        //
        // Previously an empty seed field meant "no --fingerprint flag", relying
        // on the wrapper's default args to supply a random one. The Hub launches
        // with StealthArgs=false (so it can drop the wrapper's hardcoded
        // --no-sandbox), which means those defaults no longer apply: omitting
        // the flag here would launch a browser with *no fingerprint spoofing at
        // all* while the UI still described the profile as protected.
        // ------------------------------------------------------------------
        var seed = fp.Seed is > 0 ? fp.Seed.Value : SeedFromId(profile.Id);
        SetInt("--fingerprint", seed);
        Set("--fingerprint-platform", PlatformFlag(fp.Platform));

        if (!string.IsNullOrWhiteSpace(fp.PlatformVersion))
            Set("--fingerprint-platform-version", fp.PlatformVersion);
        if (fp.Brand != BrowserBrand.Chrome)
            Set("--fingerprint-brand", fp.Brand.ToString());
        if (!string.IsNullOrWhiteSpace(fp.BrandVersion))
            Set("--fingerprint-brand-version", fp.BrandVersion);

        if (fp.Screen.Mode == ValueMode.Manual && fp.Screen.Width is > 0 && fp.Screen.Height is > 0)
        {
            SetInt("--fingerprint-screen-width", fp.Screen.Width.Value);
            SetInt("--fingerprint-screen-height", fp.Screen.Height.Value);
        }

        if (fp.Gpu.Mode == ValueMode.Manual)
        {
            if (!string.IsNullOrWhiteSpace(fp.Gpu.Vendor))
                Set("--fingerprint-gpu-vendor", fp.Gpu.Vendor);
            if (!string.IsNullOrWhiteSpace(fp.Gpu.Renderer))
                Set("--fingerprint-gpu-renderer", fp.Gpu.Renderer);
        }

        if (fp.CpuCores.Mode == ValueMode.Manual && fp.CpuCores.Value is > 0)
            SetInt("--fingerprint-hardware-concurrency", fp.CpuCores.Value.Value);
        if (fp.DeviceMemory.Mode == ValueMode.Manual && fp.DeviceMemory.Value is > 0)
            SetInt("--fingerprint-device-memory", fp.DeviceMemory.Value.Value);
        if (fp.StorageQuotaMb is > 0)
            SetInt("--fingerprint-storage-quota", fp.StorageQuotaMb.Value);
        if (fp.TaskbarHeight is >= 0)
            SetInt("--fingerprint-taskbar-height", fp.TaskbarHeight.Value);

        // Noise is ON in the binary by default; only the opt-out needs a flag.
        // The four per-surface settings collapse here because the binary has a
        // single switch — see NoiseConfig.Resolve.
        if (!fp.Noise.Resolve())
            Set("--fingerprint-noise", "false");

        if (fp.WindowsFontMetrics && fp.Platform == FingerprintPlatform.Windows)
            Set("--fingerprint-windows-font-metrics");
        if (!string.IsNullOrWhiteSpace(fp.FontsDir))
            Set("--fingerprint-fonts-dir", fp.FontsDir);
        if (fp.AllowThirdPartyCookies)
            Set("--fingerprint-allow-3p-cookies");

        // WebRTC: 'auto' is only meaningful behind a proxy — a spoofed ICE IP on
        // a direct connection is itself a mismatch, so it is skipped without one.
        if (fp.WebRtc.Mode == WebRtcMode.Manual && !string.IsNullOrWhiteSpace(fp.WebRtc.Ip))
            Set("--fingerprint-webrtc-ip", fp.WebRtc.Ip);
        else if (fp.WebRtc.Mode == WebRtcMode.Auto && profile.Proxy.IsConfigured)
            Set("--fingerprint-webrtc-ip", "auto");

        // Geolocation: explicit coordinates only. Ip mode is handled by geoip in
        // the wrapper; Off leaves the binary default untouched.
        if (profile.Geo.Mode == GeoMode.Manual && profile.Geo.Latitude is not null && profile.Geo.Longitude is not null)
        {
            var lat = profile.Geo.Latitude.Value.ToString(CultureInfo.InvariantCulture);
            var lon = profile.Geo.Longitude.Value.ToString(CultureInfo.InvariantCulture);
            Set("--fingerprint-location", $"{lat},{lon}");
        }

        // A pinned locale must also reach --lang so Accept-Language matches.
        if (profile.Locale.Mode == LocaleMode.Manual)
        {
            if (!string.IsNullOrWhiteSpace(profile.Locale.Locale))
            {
                Set("--lang", profile.Locale.Locale);
                Set("--fingerprint-locale", profile.Locale.Locale);
            }
            if (!string.IsNullOrWhiteSpace(profile.Locale.Timezone))
                Set("--fingerprint-timezone", profile.Locale.Timezone);
        }

        // User extra args last: they may add flags the app does not model, but
        // they may not hijack the identity flags above.
        foreach (var raw in profile.Startup.ExtraArgs)
        {
            var arg = raw.Trim();
            if (!arg.StartsWith("--", StringComparison.Ordinal)) continue;
            var key = arg.Split('=')[0];
            var owned = OwnedPrefixes.Any(p =>
                key == p || key.StartsWith(p + "-", StringComparison.Ordinal));
            if (owned && flags.ContainsKey(key)) continue;
            flags[key] = arg;
        }

        return [.. flags.Values];
    }
}
